File: plutus-core/plutus-core/bench/bitwise/Benches/Homogenous.hs
module Benches.Homogenous (
  benches
  ) where

import Data.Bits (zeroBits)
import Data.ByteString (ByteString)
import Data.ByteString qualified as BS
import Data.ByteString.Unsafe (unsafeUseAsCStringLen)
import Data.Word (Word8)
import DataGen (mkHomogenousArg, noCleanup, sizes)
import Foreign.C.Types (CBool (CBool), CSize (CSize), CUChar)
import Foreign.Ptr (Ptr, castPtr)
import Test.Tasty (withResource)
import Test.Tasty.Bench (Benchmark, bcompare, bench, bgroup, nfIO)

benches :: Benchmark
benches = bgroup "Basic homogeneity" $ benchBasic "Basic homogeneity" <$> sizes

-- Helpers

benchBasic ::
  String ->
  Int ->
  Benchmark
benchBasic mainLabel len =
  withResource (mkHomogenousArg len zeroBits) noCleanup $ \xs ->
    let naiveLabel = "all"
        cnaiveLabel = "naive C"
        cslidingLabel = "sliding window C"
        testLabel = mainLabel <> ", length " <> show len
        matchLabel = "$NF == \"" <> naiveLabel <> "\" && $(NF - 1) == \"" <> testLabel <> "\"" in
      bgroup testLabel [
        bench naiveLabel . nfIO $ BS.all (== zeroBits) <$> xs,
        bcompare matchLabel . bench cnaiveLabel . nfIO $ xs >>= wrapping (cHomogenousNaive zeroBits),
        bcompare matchLabel . bench cslidingLabel . nfIO $ xs >>= wrapping (cHomogenousSlidingWindow zeroBits)
        ]

-- Avoids having to rewrap C ops tediously each time
wrapping :: (Ptr CUChar -> CSize -> CBool) -> ByteString -> IO Bool
wrapping f bs = unsafeUseAsCStringLen bs $ \(ptr, len) -> do
  let (CBool res) = f (castPtr ptr) (fromIntegral len)
  pure $ res /= 0

foreign import ccall unsafe "cbits.h c_homogenous_naive"
  cHomogenousNaive ::
    Word8 ->
    Ptr CUChar ->
    CSize ->
    CBool

foreign import ccall unsafe "cbits.h c_homogenous_sliding_window"
  cHomogenousSlidingWindow ::
    Word8 ->
    Ptr CUChar ->
    CSize ->
    CBool

File: plutus-core/plutus-core/bench/bitwise/cbits/homogenous.c
#include "cbits.h"
#include <string.h>

bool c_homogenous_naive(unsigned char const needle,
                        unsigned char const *restrict haystack,
                        size_t const len) {
  bool homogenous = true;
  for (size_t i = 0; i < len; i++) {
    if (haystack[i] != needle) {
      homogenous = false;
      break;
    }
  }
  return homogenous;
}

bool c_homogenous_sliding_window(unsigned char const needle,
                                 unsigned char const *restrict haystack,
                                 size_t len) {
  for (size_t i = 0; i < 16; i++) {
    if (len == 0) {
      return true;
    }
    if (haystack[0] != needle) {
      return false;
    }
    haystack++;
    len--;
  }
  return (memcmp(haystack - 16, haystack, len) == 0);
}
